src/borg/compress.pyx

```diff
@@ -96,6 +99,51 @@ cdef class CompressorBase:
         # strip ID bytes
         return data[2:]
 
+cdef class DecidingCompressor(CompressorBase):
+    """
+    base class for (de)compression classes that ,
+    handles decide and compression based on an internal
+    _decide() method.
+    """
+    name = 'decidebaseclass'
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def _decide(self, data):
+        """
+        Decides what to do with *data*. Returns (compressor, compressed_data).
+
+        *compressed_data* can be the result of *data* being processed by *compressor*,
+        if that is generated as a side-effect of the decision process, or None otherwise.
+
+        This private method allows for more efficient implementation of compress()
+        and decide_compress() making use of *compressed_data*, if already generated.
+        """
+        return self, data
```

Comment:
hmm, isn't that something that **must** be implemented by derived classes?

so, `raise NotImplementedError` rather?

Reply:
It would work if it wasn't (it does return a compressor and data and due to the compressor being self, decide_compress() passes things up in hierarchy to eventually have an ID added), but there wouldn't be any compression, so it does not make sense to not implement it.
 (In other words, a FooCompressor derivedfrom DecidingCompressor that does not implement any methods of its own (and only sets ID) would be an overcomplicated NONE compressor).
Still go for `NotImplementedError`?

Reply:
Well, if it doesn't make sense to not override it, i think it is better to require overriding it (by raising NotImplementedError).